ruleset.go

```diff
@@ -24,15 +28,151 @@ const (
 	ErrorNoRoomForSnake  = RulesetError("not enough space to place snake")
 	ErrorNoRoomForFood   = RulesetError("not enough space to place food")
 	ErrorNoMoveFound     = RulesetError("move not provided for snake")
 	ErrorZeroLengthSnake = RulesetError("snake is length zero")
+
+	// Ruleset / game type names
+	GameTypeConstrictor = "constrictor"
+	GameTypeRoyale      = "royale"
+	GameTypeSolo        = "solo"
+	GameTypeSquad       = "squad"
+	GameTypeStanadard   = "standard"
+	GameTypeWrapped     = "wrapped"
+
+	// Game creation parameter names
+	ParamGameType            = "name"
+	ParamFoodSpawnChance     = "foodSpawnChance"
+	ParamMinimumFood         = "minimumFood"
+	ParamHazardDamagePerTurn = "damagePerTurn"
+	ParamShrinkEveryNTurns   = "shrinkEveryNTurns"
+	ParamAllowBodyCollisions = "allowBodyCollisions"
+	ParamSharedElimination   = "sharedElimination"
+	ParamSharedHealth        = "sharedHealth"
+	ParamSharedLength        = "sharedLength"
 )
 
+type rulesetBuilder struct {
+	params map[string]string // game customisation parameters
+	seed   int64             // used for random events in games
+	squads map[string]string // Snake ID -> Squad Name
+}
+
+// NewRulesetBuilder returns an instance of a builder for the Ruleset types.
+func NewRulesetBuilder() *rulesetBuilder {
+	return &rulesetBuilder{
+		params: map[string]string{},
+		squads: map[string]string{},
+	}
+}
+
+// WithParams accepts a map of game parameters for customizing games.
+//
+// Parameters are copied. If called multiple times, parameters are merged such that:
+//   - existing keys in both maps get overwritten by the new ones
+//   - existing keys not present in the new map will be retained
+//   - non-existing keys only in the new map will be added
+//
+// Unrecognised parameters will be ignored and default values will be used.
+// Invalid parameters (i.e. a non-numerical value where one is expected), will be ignored
+// and default values will be used.
+func (rb *rulesetBuilder) WithParams(params map[string]string) *rulesetBuilder {
+	for k, v := range params {
+		rb.params[k] = v
+	}
+	return rb
+}
+
+// WithSeed sets the seed used for randomisation by certain game modes.
+func (rb *rulesetBuilder) WithSeed(seed int64) *rulesetBuilder {
+	rb.seed = seed
+	return rb
+}
+
+// AddSnakeToSquad adds the specified snake (by ID) to a squad with the given name.
+// This configuration may be ignored by game modes if they do not support squads.
+func (rb *rulesetBuilder) AddSnakeToSquad(snakeID, squadName string) *rulesetBuilder {
+	rb.squads[snakeID] = squadName
+	return rb
+}
+
+// Ruleset constructs a customised ruleset using the parameters passed to the builder.
+func (rb rulesetBuilder) Ruleset() Ruleset {
+	standardRuleset := &StandardRuleset{
+		FoodSpawnChance:     paramsInt32(rb.params, ParamFoodSpawnChance, 0),
+		MinimumFood:         paramsInt32(rb.params, ParamMinimumFood, 0),
+		HazardDamagePerTurn: paramsInt32(rb.params, ParamHazardDamagePerTurn, 0),
+	}
```

Comment:
I think we need some way of passing along `hazardMap` and `hazardMapAuthor` here, or they'll get lost and won't be sent along to the snakes as currently implemented.

One way could be adding them as optional fields on StandardRuleset and returning them back out in the Settings method, even if they're not used? Seems kind of awkward.

The other option I can think of is engine adding these fields back in manually when making the requests. That doesn't look too difficult to do. 

What do you think?

Reply:
Good catch - I totally forgot about `hazardMap` and `hazardMapAuthor` after merging them in, since they don't seem to be used by anything in the rulesets.

I'll look at passing them through the standard ruleset like you suggested.

Reply:
This is done - I went with your first suggestion: adding them to standard mode.

I included adding new params for `hazardMap` and `hazardMapAuthor`.

Annnd I also added integration tests (using the CLI Play command) to ensure that the settings get serialized and make sure the new `hazardMap` and `hazardMapAuthor` get passed through. Each of the existing 6 game modes has one of these tests. It's nice because it gives a pretty good idea that things are working E2E from param parsing all the way to serialization.